### PR TITLE
[CPDLP-3652] Add unique index to all models with ecf id in NPQ

### PR DESCRIPTION
**Pull request by leandroalemao**
### Context

Ticket: [CPDLP-3652](https://dfedigital.atlassian.net/browse/CPDLP-3652)

We currently do not have a unique index on models with `ecf_id` in NPQ. We will need to add those except for `Users` which needs fixing and is being done by the NPQ team.

### Changes proposed in this pull request

- Add a unique index on `ecf_id` all models with an `ecf_id` in NPQ (except for User) as well as a DB constraint;
- Ensure there is no data in all other envs: staging/sandbox/separation/production which will break this index, and fix the data;
- Ensure the indices added will run quickly and not block deployments;

Tested in snapshot db:

```
== 20241022204916 AddUniqueConstraintToEcfIdInApplications: migrating =========
-- remove_index(:applications, :ecf_id)
   -> 0.2731s
-- add_index(:applications, :ecf_id, {:unique=>true, :algorithm=>:concurrently})
   -> 0.4602s
== 20241022204916 AddUniqueConstraintToEcfIdInApplications: migrated (0.7378s) 

== 20241023093413 AddUniqueConstraintToEcfIdInContractTemplates: migrating ====
-- remove_index(:contract_templates, :ecf_id)
   -> 0.0660s
-- add_index(:contract_templates, :ecf_id, {:unique=>true, :algorithm=>:concurrently})
   -> 0.0399s
== 20241023093413 AddUniqueConstraintToEcfIdInContractTemplates: migrated (0.1065s) 

== 20241023093913 AddUniqueConstraintToEcfIdInCourses: migrating ==============
-- add_index(:courses, :ecf_id, {:unique=>true, :algorithm=>:concurrently})
   -> 0.0331s
== 20241023093913 AddUniqueConstraintToEcfIdInCourses: migrated (0.0334s) =====

== 20241023125017 AddUniqueConstraintToEcfIdInDeclarations: migrating =========
-- remove_index(:declarations, :ecf_id)
   -> 0.1625s
-- add_index(:declarations, :ecf_id, {:unique=>true, :algorithm=>:concurrently})
   -> 0.0510s
== 20241023125017 AddUniqueConstraintToEcfIdInDeclarations: migrated (0.2171s) 

== 20241023141116 AddUniqueConstraintToEcfIdInLeadProviders: migrating ========
-- add_index(:lead_providers, :ecf_id, {:unique=>true, :algorithm=>:concurrently})
   -> 0.0357s
== 20241023141116 AddUniqueConstraintToEcfIdInLeadProviders: migrated (0.0362s) 

== 20241023141630 AddUniqueConstraintToEcfIdInParticipantIdChanges: migrating =
-- remove_index(:participant_id_changes, :ecf_id)
   -> 0.1510s
-- add_index(:participant_id_changes, :ecf_id, {:unique=>true, :algorithm=>:concurrently})
   -> 0.0367s
== 20241023141630 AddUniqueConstraintToEcfIdInParticipantIdChanges: migrated (0.1898s)
```

Tested in the remaining envs:

```
=> [Application, ContractTemplate, Course, Declaration, LeadProvider, ParticipantIdChange].map do |klass|
  klass.where.not(ecf_id: nil).select(:ecf_id).group(:ecf_id).having("count(*) > 1").size
end
=> [{}, {}, {}, {}, {}, {}]
```
[CPDLP-3652]: https://dfedigital.atlassian.net/browse/CPDLP-3652?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by cwrw — COMMENTED** [review]
could we check data on all envs if they pass, except production as you already tested :)

File: spec/models/application_spec.rb [hunk 1]
```diff
@@ -19,6 +19,8 @@
   end
 
   describe "validations" do
+    it { is_expected.to validate_uniqueness_of(:ecf_id).case_insensitive.with_message("ECF ID must be unique").allow_nil }
```
Comment (cwrw) [inline spec/models/application_spec.rb:22]: can we add a test for the feature flag as well?

File: app/models/application_state.rb [hunk 2]
```diff
@@ -4,6 +4,9 @@ class ApplicationState < ApplicationRecord
   belongs_to :application
   belongs_to :lead_provider, optional: true
 
+  # TODO: remove "allow_nil" and add default value "gen_random_uuid()" and constraints into the DB after separation
```
Comment (cwrw) [inline app/models/application_state.rb:7]: this isn't surface in the API, so can allow it to remain `nil` after separation and delete that field as well

File: app/models/cohort.rb [hunk 3]
```diff
@@ -15,6 +15,9 @@ class Cohort < ApplicationRecord
               message: "Choose true or false for funding cap",
             }
 
+  # TODO: remove "allow_nil" and add default value "gen_random_uuid()" and constraints into the DB after separation
```
Comment (cwrw) [inline app/models/cohort.rb:18]: same comment as above, we left those intentionally not auto generated in the db: the ones that aren't surfaced in the API

File: app/models/participant_outcome.rb [hunk 4]
```diff
@@ -2,7 +2,7 @@ class ParticipantOutcome < ApplicationRecord
   belongs_to :declaration
   has_many :participant_outcome_api_requests
 
-  validates :ecf_id, uniqueness: true
+  validates :ecf_id, uniqueness: { case_sensitive: false }
```
Comment (TBenedyk) [inline app/models/participant_outcome.rb:5]: Why are we making all these case insensitive?

Comment (leandroalemao) [reply app/models/participant_outcome.rb:5]: Unique constraints are `case insensitive` or `case_sensitive false` by default so the DB won't allow adding a value for ecf_id = "583162EE-00B8-4AB9-A668-996E8EAAD40B" if "583162ee-00b8-4ab9-a668-996e8eaad40b" is already there. 

Shoulda matches in the other hand[ is `case sensitive true` by default](https://github.com/thoughtbot/shoulda-matchers/blob/main/lib/shoulda/matchers/active_record/validate_uniqueness_of_matcher.rb#L190), so we need to add a `.case_insensitive` to the matcher, so the test passes.

`it { is_expected.to validate_uniqueness_of(:ecf_id).case_insensitive.with_message("ECF ID must be unique") }`

As we had to add `case_insensitive` to the spec, I'm adding `{ case_sensitive: false }` into the model as well to keep it consistent with the rest.

**Review by mooktakim — APPROVED** [review]

**Review by ethax-ross — APPROVED** [review]

**Pull request merged**